File: lib/agent-kernel/budget.ts
import type { RunBudget } from './types'
import type { Blackboard } from './blackboard'

/**
 * Vérifie le budget d'un run et retourne une raison lisible si dépassé,
 * ou null si on peut continuer. Utilisé pour garder la boucle
 * réflexion → décision bornée et prévisible (coût, latence, timeout Vercel)
 * plutôt que de laisser les agents boucler tant qu'ils "jugent nécessaire".
 */
export function budgetExceeded(budget: RunBudget, blackboard: Blackboard): string | null {
  if (blackboard.modelCallCount >= budget.maxModelCalls) {
    return `budget d'appels modèle atteint (${budget.maxModelCalls})`
  }
  if (Date.now() - budget.startedAt >= budget.deadlineMs) {
    return `deadline dépassée (${Math.round(budget.deadlineMs / 1000)}s)`
  }
  return null
}

export function createBudget(overrides: Partial<Omit<RunBudget, 'startedAt'>> = {}): RunBudget {
  return {
    maxIterations: 3,
    maxModelCalls: 12,
    deadlineMs: 260_000,
    ...overrides,
    startedAt: Date.now(),
  }
}

File: backend/agent_kernel/budget.py
"""Port de lib/agent-kernel/budget.ts.

Vérifie le budget d'un run et retourne une raison lisible si dépassé, ou
None si on peut continuer. Garde la boucle réflexion → décision bornée et
prévisible (coût, latence) plutôt que de laisser les agents boucler tant
qu'ils "jugent nécessaire".
"""
import time

from .blackboard import Blackboard
from .types import RunBudget


def budget_exceeded(budget: RunBudget, blackboard: Blackboard) -> str | None:
    if blackboard.model_call_count >= budget.max_model_calls:
        return f"budget d'appels modèle atteint ({budget.max_model_calls})"
    if time.time() - budget.started_at >= budget.deadline_seconds:
        return f"deadline dépassée ({round(budget.deadline_seconds)}s)"
    return None


def create_budget(
    max_iterations: int = 3,
    max_model_calls: int = 12,
    deadline_seconds: float = 260.0,
) -> RunBudget:
    return RunBudget(
        max_iterations=max_iterations,
        max_model_calls=max_model_calls,
        deadline_seconds=deadline_seconds,
        started_at=time.time(),
    )
